python/pylibraft/pylibraft/neighbors/cagra/cagra.pyx

```diff
@@ -0,0 +1,852 @@
+            handle = DeviceResources()
+        cdef device_resources* handle_ = \
+            <device_resources*><size_t>handle.getHandle()
+
+        self.index = new c_cagra.index[int8_t, uint32_t](
+            deref(handle_))
+
+    def __repr__(self):
+        m_str = "metric=" + _get_metric_string(self.index.metric())
+        attr_str = [attr + "=" + str(getattr(self, attr))
+                    for attr in ["metric", "dim", "graph_degree"]]
+        attr_str = m_str + attr_str
+        return "Index(type=CAGRA, " + (", ".join(attr_str)) + ")"
+
+    @property
+    def metric(self):
+        return self.index[0].metric()
+
+    @property
+    def size(self):
+        return self.index[0].size()
+
+    @property
+    def dim(self):
+        return self.index[0].dim()
+
+    @property
+    def graph_degree(self):
+        return self.index[0].graph_degree()
+
+    def __dealloc__(self):
+        if self.index is not NULL:
+            del self.index
+
+
+cdef class IndexUint8(Index):
+    cdef c_cagra.index[uint8_t, uint32_t] * index
+
+    def __cinit__(self, handle=None):
+        if handle is None:
+            handle = DeviceResources()
+        cdef device_resources* handle_ = \
+            <device_resources*><size_t>handle.getHandle()
+
+        self.index = new c_cagra.index[uint8_t, uint32_t](
+            deref(handle_))
+
+    def __repr__(self):
+        m_str = "metric=" + _get_metric_string(self.index.metric())
+        attr_str = [attr + "=" + str(getattr(self, attr))
+                    for attr in ["metric", "dim", "graph_degree"]]
+        attr_str = m_str + attr_str
+        return "Index(type=CAGRA, " + (", ".join(attr_str)) + ")"
+
+    @property
+    def metric(self):
+        return self.index[0].metric()
+
+    @property
+    def size(self):
+        return self.index[0].size()
+
+    @property
+    def dim(self):
+        return self.index[0].dim()
+
+    @property
+    def graph_degree(self):
+        return self.index[0].graph_degree()
+
+    def __dealloc__(self):
+        if self.index is not NULL:
+            del self.index
+
+
+@auto_sync_handle
+@auto_convert_output
+def build(IndexParams index_params, dataset, handle=None):
+    """
+    Build the CAGRA index from the dataset for efficient search.
+
+    The build consist of two steps: build an intermediate knn-graph, and
+    optimize it to create the final graph. The index_params object controls
+    the node degree of these graphs.
+
+    It is required that dataset and the optimized graph fit the GPU memory.
+
+    To customize the parameters for knn-graph building and pruning, and to
+    reus the intermediate results, you could build the index in two steps
+    using pylibraft.cagra.build_knn_graph and
+    pylibraft.cagra.optimize.
+
+     The following distance metrics are supported:
+     - L2
+
+    Parameters
+    ----------
+    index_params : IndexParams object
+    dataset : array interface compliant matrix shape (n_samples, dim)
```

Comment:
Can the dataset also exist on the GPU? Chould it support both cuda array interface AND array interface?